server/src/main/java/org/elasticsearch/search/vectors/ChildBlockJoinVectorScorerProvider.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.ConjunctionUtils;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.HitQueue;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.join.BitSetProducer;
+import org.apache.lucene.search.join.DiversifyingChildrenByteKnnVectorQuery;
+import org.apache.lucene.search.join.DiversifyingChildrenFloatKnnVectorQuery;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.core.Nullable;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * This class is used to score vector child documents given the results of a {@link DiversifyingChildrenByteKnnVectorQuery}
+ * or {@link DiversifyingChildrenFloatKnnVectorQuery}.
+ * It iterates the previously matched children, and allows for scoring and finding `numChildrenPerParent` for each parent.
+ */
+class ChildBlockJoinVectorScorerProvider {
```

Comment:
This isn't a true `weight` and honestly shouldn't satisfy that interface. Thus I opted for a simpler but custom interface here.